File: python/Problem1.py
class Solution:
    def getMaxProfit(self, prices: list[int]) -> int:
        n = len(prices)
        if (n < 2): return 0

        left = [0] * n
        min_price = prices[0]

        for i in range(1, n):
            min_price = min(min_price, prices[i])
            left[i] = max(left[i-1], prices[i] - min_price)

        right = [0] * n
        max_price = prices[n-1]

        for i in range(n-2, -1, -1):
            max_price = max(max_price, prices[i])
            right[i] = max(right[i+1], max_price - prices[i])

        max_profit = 0
        for i in range(n):
            max_profit = max(max_profit, left[i] + (right[i+1] if i+1 < n else 0))

        return max_profit
    
solution = Solution()

input1 = [3,3,5,0,0,3,1,4]
result1 = solution.getMaxProfit(input1)
print(result1)

input2 = [1,2,3,4,5]
result2 = solution.getMaxProfit(input2)
print(result2)

input3 = [7,6,4,3,1]
result3 = solution.getMaxProfit(input3)
print(result3)

File: TypeScript/Problem1.ts
function getMaxProfit(prices: number[]): number {
    const n = prices.length;
    if (n < 2) return 0;
    
    const left = new Array(n).fill(0);
    let minPrice = prices[0];
    
    for (let i = 1; i < n; i++) {
        minPrice = Math.min(minPrice, prices[i]);
        left[i] = Math.max(left[i-1], prices[i] - minPrice);
    }
    
    const right = new Array(n).fill(0);
    let maxPrice = prices[n-1];
    
    for (let i = n-2; i >= 0; i--) {
        maxPrice = Math.max(maxPrice, prices[i]);
        right[i] = Math.max(right[i+1], maxPrice - prices[i]);
    }
    
    let maxProfit = 0;
    for (let i = 0; i < n; i++) {
        maxProfit = Math.max(maxProfit, left[i] + (i+1 < n ? right[i+1] : 0));
    }
    
    return maxProfit;
};

const input1 = [3,3,5,0,0,3,1,4];
const result1 = getMaxProfit(input1);
console.log(result1);

const input2 = [1,2,3,4,5];
const result2 = getMaxProfit(input2);
console.log(result2);

const input3 = [7,6,4,3,1];
const result3 = getMaxProfit(input3);
console.log(result3);